pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+    pub fn do_charge_payment_stream(
+        bsp_account: &T::AccountId,
+        user_account: &T::AccountId,
+    ) -> Result<BalanceOf<T>, DispatchError> {
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Check that a payment stream between that BSP and user exists
+        ensure!(
+            PaymentStreams::<T>::contains_key(bsp_id, user_account),
+            Error::<T>::PaymentStreamNotFound
+        );
+
+        // Get the information of the payment stream
+        let payment_stream = PaymentStreams::<T>::get(bsp_id, user_account)
+            .ok_or(Error::<T>::PaymentStreamNotFound)?;
+
+        // Note: No need to check if a new proof has been submitted since the last charge as the only consecuence of that is charging 0 to the user,
+        // and not erroring out helps to be able to call this function without errors when updating or removing a payment stream.
+
+        // Calculate the time passed between the last valid proof and the last charged proof
+        let time_passed = expect_or_err!(payment_stream
+            .last_valid_proof
+            .checked_sub(&payment_stream.last_charged_proof), "Last valid proof should always be greater than or equal to the last charged proof, inconsistency error.",
+            Error::<T>::LastChargeGreaterThanLastValidProof);
+
+        // Convert it to the balance type (for math)
+        let time_passed_balance_typed = T::BlockNumberToBalance::convert(time_passed);
+
+        // Calculate the amount to charge
+        let amount_to_charge = payment_stream
+            .rate
+            .checked_mul(&time_passed_balance_typed)
+            .ok_or(Error::<T>::ChargeOverflow)?;
+
+        // Check the free balance of the user
+        let user_balance = T::NativeBalance::reducible_balance(
+            &user_account,
+            Preservation::Preserve,
+            Fortitude::Polite,
+        );
+
+        // If the user does not have enough balance to pay for its storage:
+        if user_balance < amount_to_charge {
+            // Flag it in the UsersWithoutFunds mapping and emit the UserWithoutFunds event
+            UsersWithoutFunds::<T>::insert(user_account, ());
+            Self::deposit_event(Event::<T>::UserWithoutFunds {
+                who: user_account.clone(),
+            });
+
+            // Return 0 as the amount that has been charged
+            Ok(Zero::zero())
+        } else {
+            // If the user does have enough funds to pay for its storage:
+
+            // Clear the user from the UsersWithoutFunds mapping
+            // TODO: Design a more robust way of handling out-of-funds users
+            UsersWithoutFunds::<T>::remove(user_account);
```

Comment:
Shoulnd't you check first if the user is in the mapping?

Reply:
Should I? If it's not on the map, the `remove` method should do nothing, so we don't lose anything by not checking first if we have to remove it either way